src/flamegraph/color.rs

```diff
@@ -0,0 +1,272 @@
+        Palette::Mem => return ((0, 0, v3), (190, 50, v2), (0, 210, v1)),
+        Palette::Io => return ((80, 60, v1), (80, 60, v1), (190, 55, v2)),
+        Palette::Red => return ((200, 55, v1), (50, 80, v1), (50, 80, v1)),
+        Palette::Green => return ((50, 60, v1), (200, 55, v1), (50, 60, v1)),
+        Palette::Blue => return ((80, 60, v1), (80, 60, v1), (205, 60, v1)),
+        Palette::Yellow => return ((175, 55, v1), (175, 55, v1), (50, 20, v1)),
+        Palette::Purple => return ((190, 65, v1), (80, 60, v1), (190, 65, v1)),
+        Palette::Aqua => return ((50, 60, v1), (165, 55, v1), (165, 55, v1)),
+        Palette::Orange => return ((190, 65, v1), (90, 65, v1), (0, 0, v1)),
+        Palette::Java => handle_java_palette(name),
+        Palette::Perl => handle_perl_palette(name),
+        Palette::Js => handle_js_palette(name),
+        Palette::Wakeup => handle_wakeup_palette(name),
+        Palette::Chain => handle_chain_palette(name),
+    };
+
+    coefficients_for_palette(&real_palette, name, v1, v2, v3)
+}
+
+fn affine_transform(a: u8, b: u8, x: f32) -> u8 {
+    b + (a as f32 * x) as u8
+}
+
+fn color_from_palette(palette: &Palette, name: &str, v1: f32, v2: f32, v3: f32) -> String {
+    let ((r_b, r_a, r_x),
+         (g_b, g_a, g_x),
+         (b_b, b_a, b_x)) = coefficients_for_palette(palette, name, v1, v2, v3);
+
+    let red = affine_transform(r_a, r_b, r_x);
+    let green = affine_transform(g_a, g_b, g_x);
+    let blue = affine_transform(b_a, b_b, b_x);
+
+    format!("rgb({},{},{})", red, green, blue)
+}
+
+pub(super) fn color_map<'a>(palette: &Palette, hash: bool, name: &'a str, palette_map: &'a mut HashMap<String, String>) -> &'a str {
+    palette_map.entry(name.to_string()).or_insert_with(|| color(palette, hash, name))
+}
+
+pub(super) fn color(palette: &Palette, hash: bool, name: &str) -> String {
+    let (v1, v2, v3) = if hash {
+        let name_hash = namehash(name);
+        let reverse_name_hash = namehash(&name.chars().rev().collect::<String>());
+
+        (name_hash, reverse_name_hash, reverse_name_hash)
+    } else {
+        (rand::random(), rand::random(), rand::random())
+    };
+
+    color_from_palette(palette, name, v1, v2, v3)
+}
+
+pub(super) fn get_background_colors_for(palette: &Palette) -> (&str, &str) {
+    match palette {
+        Palette::Hot | Palette::Java | Palette::Js | Palette::Perl => YELLOW_GRADIENT,
+        Palette::Mem | Palette::Chain => BLUE_GRADIENT,
+        _ => GRAY_GRADIENT
+    }
+}
+
+pub(super) fn read_palette(file: &str) -> Result<HashMap<String, String>, io::Error> {
+    let mut map = HashMap::default();
+    let path = Path::new(file);
+
+    if path.exists() {
+        let file = File::open(path)?;
+        let file = BufReader::new(file);
+
+        for line in file.lines() {
+            let line = line?;
+            let words = line.split("->").collect::<Vec<_>>();
+            map.insert(words[0].to_string(), words[1].to_string());
+        }
+    }
```

Comment:
A comment about why the `else` case isn't an error would be good here.

Reply:
[Done](https://github.com/jonhoo/inferno/pull/32/commits/d5000681e16a220684e72429b32d5115588c483a)